src/calendar_duration.rs

```diff
@@ -0,0 +1,244 @@
+use core::num::NonZeroU32;
+
+use crate::{expect, try_opt};
+
+/// ISO 8601 duration type.
+///
+/// With this type a duration can be expressed as a combination of multiple components: months,
+/// days, minutes, seconds and nanoseconds.
+///
+/// # Examples
+///
+/// ```
+/// # use chrono::CalendarDuration;
+/// // Encoding 1½ year as a duration of a year and 6 months:
+/// let _duration = CalendarDuration::new().years_and_months(1, 6).unwrap();
+///
+/// // Encoding 1½ year as a duration of 12 months and 183 days (366 / 2):
+/// let _duration = CalendarDuration::new().months(12).days(183);
+///
+/// // Encoding 1½ year as a duration of 548 days (365 + 366 / 2):
+/// let _duration = CalendarDuration::new().days(548);
+///
+/// // Encoding 1½ year as a duration in seconds:
+/// let _duration = CalendarDuration::new().seconds(548 * 24 * 60 * 60);
+/// ```
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct CalendarDuration {
+    // Components with a nominal duration
+    months: u32,
+    days: u32,
+    // Components with an accurate duration
+    // `seconds` can either encode `minutes << 6 | seconds`, or just seconds.
+    seconds: u32,
+    // `nanos` encodes `nanoseconds << 2 | has_minutes << 1 | 1`.
+    nanos: NonZeroU32,
```

Comment:
I'd prefer to implement the boring version here and optimize memory usage later if we decide that is necessary. Since these are public fields anyway, there should be no need to prematurely optimize.

Reply:
There is a little premature optimization going on. But my goal is more about having two clear modes of operation. My thought process went like this:

1. Include a flag for 'I care about leap seconds'/'leap seconds don't exist'.
   However that is not really something we can serialize, and also not enough to make `CalendarDuration` at least in theory capable of working with leap seconds correctly.
2. Have explicit `minutes` and `seconds` fields.
3. Suppose we pre-multiply the `minutes` field with `60`. This turns the fields into `seconds_in_utc` and `seconds_in_tai`.
   This is becoming a mess. Having two pretty much the same fields that only differ in how they count leap seconds within one duration seems unreasonable.
4. Introduce two sane modes of operation:
   - arbitrary large minutes and only sub-minute seconds
   - arbitrary large seconds

To encode that with rust types:
```rust
pub struct CalendarDuration {
    // Components with a nominal duration
    months: u32,
    days: u32,
    // Components with an accurate duration
    accurate: AccurateDuration,
}

enum AccurateDuration {
    MinutesAndSubMinuteSecondsAndNanos(u32, u8, u32),
    SecondsAndNanos(u32, u32);
}
```

We can simplify that a bit if we manually uphold the relation between `minutes` and `seconds` fields to:
```rust
pub struct CalendarDuration {
    // Components with a nominal duration
    months: u32,
    days: u32,
    // Components with an accurate duration
    minutes: u32,
    seconds: u32,
    nanos: u32,
}
```

The idea is that if you have an existing `CalendarDuration` and change it with `with_seconds`, it should switch to the `SecondsAndNanos` mode. Previous minutes and seconds should be forgotten. And if `minutes` is non-zero, `seconds` must be less than `61`.

If we already have to uphold a relation between `minutes` and `seconds` anyway, why not encode them in a single `u32` and get the smaller type size for free? And if we don't do it now but want to keep the option for the optimization open for later, we should already include the range checks.

I have to admit stuffing the discriminant inside the `nanos` field, and using the other bit for niche optimization, is a premature optimization. Can we please keep it?

Reply:
IMO that's way too much magic. I'd rather have the actual enum if we're going to do something like this.

Reply:
I added an `MinutesAndSeconds` type to encode this in.

And I switched to accurate part of a duration to a 64-bit type.
With a 32-bit value we could only store an accurate duration of ca. 136 years. With a 64-bit type `CalendarDuration` can cover all uses over `std::time::Duration` and `TimeDelta` (forgetting the negative part). With this type we also never have to make methods fallible that construct a duration from the difference between two dates; the result always fits.